assets/src/components/Lightbox/Lightbox.js

```diff
@@ -0,0 +1,154 @@
+import React, { useEffect, useRef, useState } from 'react';
+
+import PhotoSwipe from '../../../../node_modules/photoswipe/dist/photoswipe.js';
+import PhotoSwipeUI_Default from '../../../../node_modules/photoswipe/dist/photoswipe-ui-default.js';
+
+export const setupLightboxForImageBlocks = function() {
+  const images = document.querySelectorAll('.wp-block-image img');
+  const items = [...images].map(image => ({
+    src: image.src ? image.src : image.dataset.src,
+    w: 0,
+    h: 0
+  }));
+
+  // We can't use `createPortal` outside `render()`,
+  // `React.render()` needs a node, even if it ends up being empty.
+  // See: https://github.com/facebook/react/issues/12653#issuecomment-382851495
+  const lightBoxNode = document.createElement('div');
+  lightBoxNode.style.position = 'fixed';
+
+  document.body.appendChild(lightBoxNode);
+
+  const imageBlocks = document.querySelectorAll('.wp-block-image');
+  imageBlocks.forEach((imageBlock, index) => {   
+    const caption = imageBlock.querySelector('figcaption');
+    if (caption) {
+      items[index].title = caption.innerHTML;
+    }
+
+    imageBlock.querySelector('img').addEventListener('click', () => wp.element.render(
+        <Lightbox isOpen={true} index={index} items={items} onClose={() => {}} />,
+        lightBoxNode
+      )
+    );
+  })
+
+  return items;
+}
+
+export const wpImagesToItems = images => images.map(
+  image => ({
+    src: image.image_src,
+    w: 0,
+    h: 0,
+  })
+);
+
+export const Lightbox = props => {
+  let pswpElement = useRef(null);
+
+  const options = {
+    index: props.index || 0,
+    closeOnScroll: false,
+    history: false
+  };
+
+  const [currentIndex, setCurrentIndex] = useState(options.index);
+
+  useEffect(() => {
+    const photoSwipe = new PhotoSwipe(pswpElement, PhotoSwipeUI_Default, props.items, options);
```

Comment:
We don't want this to happen more than once per opened lightbox. 

Currently, when using a Carousel/Slider Gallery, this effect is ran each time the Carousel goes to the next slide.

Reply:
I think having a complex dependency is always a bad thing with effects, they should be as simple a value as possible, relating to a specific effect. So definitely passing the entire props of the component seems like a bad idea, and causes the exact situation you want to avoid, i.e. no control over when things change.

For instance here you could have a single effect that only creates the `PhotoSwipe` on the initial render. If the instantiation of `PhotoSwipe` isn't slow/expensive we could do it synchronously as well using state.
```js
const [photoSwipe] = useState(()=>(new PhotoSwipe(pswpElement, PhotoSwipeUI_Default, props.items, options))
```

Reply:
Actually since we only ever display a single lightbox, we don't need to create it in react, we could include the initial markup in the page, then use a hook on each image block that will add the image to the lightbox, and exposes a function to open the lightbox at that image.

Reply:
Hey monsieur! Not creating the HTML in React would be easy as it's just delivering the needed markup from a different place, but I see a few downsides. I guess you mean detecting whether a page/post has image or gallery blocks in the server side (probably using `has_block`) and include the HTML for the Lightbox depending on that. To include the markup in the page I guess we'll have to do that in the master theme, not in the blocks plugin. We'll probably need some condition in a Twig template and pass some variable to the Timber context. Is that right?

I think the benefit of the current approach is that the code is entirely encapsulated in the Lightbox component file instead of having logic scattered through the repos (or between PHP & JS in the same repo). There's also the advantage of having the HTML right there in case we need to make further customizations. What do you think?

Reply:
> I guess you mean detecting whether a page/post has image or gallery blocks in the server side (probably using has_block) and include the HTML for the Lightbox depending on that.

It might not even be worth checking if a post has an image, the vast majority of them will do. And it's a tiny amount of markup anyway, it doesn't include the actually markup for the images. It's just a starting point Photoswipe needs once per page load, then in JS we just need to "hydrate" it and tell it to render certain images.

> To include the markup in the page I guess we'll have to do that in the master theme, not in the blocks plugin.

We could just as well include this in the blocks plugin if we want to keep it together with the JS, I don't see any reason why we couldn't.

> We'll probably need some condition in a Twig template and pass some variable to the Timber context. Is that right?

I think we can get away with the simplicity of including it each time, as the vast majority of posts will use it anyway. But if we do want to make it conditional that seems like a very minimal change. We also don't need to pass any variables, it's 100% static markup. It doesn't even need to be a twig file, you can include any file in a twig template.

```twig
<body>
...
{% include "photoswipe-skeleton.html" %}
</body>
```

Reply:
Alright, that does make it simpler, but it'll be a bit more difficult to render some customizations as the indicators (currently rendered here: https://github.com/greenpeace/planet4-plugin-gutenberg-blocks/pull/414/files#diff-7603fc2cbf7cc1eec26221206dffa1f05c92e5b18163c0ab259a2b83443d1767R139).

Reply:
Ok, so I fixed the original issue by using `memo()` as a functional version of `shouldComponentUpdate`. It seems to be working as expected now.

Reply:
Are you sure that's a solid solution? React seems to recommend against relying on it to prevent a render.

> This method only exists as a performance optimization. Do not rely on it to “prevent” a render, as this can lead to bugs.

from https://reactjs.org/docs/react-api.html#reactmemo

Reply:
Yes, I read that while investigating `memo()`, but they don't shed much light on *why* it would cause bugs, I guess it's the same reason `shouldComponentUpdate` [should be used sparingly](https://reactjs.org/docs/react-component.html#shouldcomponentupdate). It's the same warning, and I think it's because the API may change, or you may have to update the comparison function if it's too complex and you change the props, or you can actually slow down the app if you are doing a really deep equality check in the comparison function. As this is not dealing with thousands of elements, I don't think the latter is an issue.

But if you look at the community, `shouldComponentUpdate` is still being used a lot to prevent re-renders, specially since most React examples in the docs use classes, not functional compos. I guess there's a reason it has not been deprecated or hidden behind some `__unsafe` prefix.

The problem with functional components is that they will always re-render their childs. There are a lot of community posts about using `memo()` as a workaround when the component is re-rendered using the same props ([example](https://dmitripavlutin.com/use-react-memo-wisely/)). It's pretty much the same as using `shouldComponentUpdate`, hence the same warnings apply.

So yes, I knew the warning, but I think it's safe as a workaround, the arguments against it seem too defensive to me, if I have to change the props of the Lightbox, I'll probably be changing some of its implementation too, and I can take care of any possible issues or change the implementation in that moment.

Reply:
> they don't shed much light on why it would cause bugs

Yeah unfortunately they don't go into detail in the those docs, but I wouldn't assume that is because they don't have a good reason to say that, they probably forgot or didn't care to add it the docs. But it's a pretty clear and unambiguous statement "**Do not** rely on it to prevent a render...".

They do go into a little more detail as to why you can't rely on it actually preventing the render in all cases in [the `useMemo` hook docs](https://reactjs.org/docs/hooks-reference.html#usememo).

> You may rely on useMemo as a performance optimization, not as a semantic guarantee. In the future, React may choose to “forget” some previously memoized values and recalculate them on next render, e.g. to free memory for offscreen components. **Write your code so that it still works without useMemo** — and then add it to optimize performance.

Reply:
I think there was a comment about GalleryFrontend around here... 

Anyway: as suggested, I moved the Lightbox component to GalleryFrontend and it definitely fixes the issue, no need for `memo()` then.